src/DotNet.Testcontainers/Clients/TestcontainersClient.cs

```diff
@@ -4,16 +4,20 @@ namespace DotNet.Testcontainers.Clients
   using System.Collections.Generic;
   using System.Diagnostics;
   using System.IO;
+  using System.Text;
   using System.Threading;
   using System.Threading.Tasks;
   using Docker.DotNet.Models;
   using DotNet.Testcontainers.Containers.Configurations;
   using DotNet.Testcontainers.Containers.OutputConsumers;
   using DotNet.Testcontainers.Images.Configurations;
   using DotNet.Testcontainers.Services;
+  using ICSharpCode.SharpZipLib.Tar;
 
   internal sealed class TestcontainersClient : ITestcontainersClient
   {
+    private readonly SemaphoreSlim semaphoreSlim = new SemaphoreSlim(1, 1);
```

Comment:
By introducing the `OnContainerStartedAsync` Callback it is possible to run into a deadlock, because `OnContainerStartedAsync` itself runs in a part with those semaphore locks. So when I call `ExecAsync` or `CopyFileAsync` inside `OnContainerStartedAsync` the semaphore is already locked. Therefore I thought it would be better to move the semaphore to the client as I thought that the semaphore was actually introduced because we should not call methods of the client in parallel.

Reply:
Good point 👍 but I think we need to take a closer look at this. Locking **each** individual method in the client is not enough. This will not synchronize the public methods. E. g. we don't know if anything else is called between `AttachAsync` and `StartAsync`. This applies to all public methods. Furthermore, the client is responsible for more than just 1 container.

We need to protect the public methods and the execution order. I need some time to think about a proper solution:

- Maybe we can implement it with a second semaphore.
- Do we really need to lock `ExecAsync` or `CopyFileAsync`?
    - We can provide a safe interface to the builder such as:
    ```csharp
    public interface ITestcontainersBuilder<out TDockerContainer>
      where TDockerContainer : IDockerContainer
    {
      ITestcontainersBuilder<TDockerContainer> OnContainerStarted(Action<ISafeInterfaceWithExecAndCopyFileAsync> container);
    }

    new TestcontainersBuilder<TestcontainersContainer>()
      .OnContainerStarted(container => container.ExecAsync(new [] { "" }))
    ```

Do you have another idea?

Reply:
I am not sure why it is important that `StartAsync` and `AttachAsync` are called subsequently without any other calls between them. Those other calls are always independent from these calls and can not corrupt any state, can they?

Reply:
I have implemented a small [example](https://pastebin.com/raw/0CnDJqUg) using both approaches. You can see that the results of both implementations are different. Your approach is now no longer deterministic and can lead to different results. 

I don't know if this breaks existing code or if it's just theoretical (ofc the example code force this behavior).

Reply:
Yes, generally one should be careful with concurrency, but I don't know if we even need any semaphore. Which state or resource are we protecting from corruption using this semaphore? Is the dotnet Docker client not thread-safe?

If some application depends on `StartAsync` to be finished, then it should wait for the result of the `StartAsync`-Task. Otherwise it is kind of an accident that the application works.

Reply:
I’ll do the review later that day. I just noticed one thing, shouldn’t be the configuration from `TestcontainersBuilderKafkaExtension` in `KafkaTestcontainerConfiguration` like in the other modules?

<hr />

In reference to the semaphore:
> Which state or resource are we protecting from corruption using this semaphore?
1. Each container instance itself.
2. We protect the field `this.container` in `TestcontainersContainer`.
3. We protect the developers from running into unexpected behaviors and race conditions.

We’re moving the semaphore **only** to work around a deadlock. This will probably cause other issues and make some parts unreliable. I don’t think it’s worth it without considering other solutions.

To get this pull request done, I suggest following:
- We keep the semaphore in `TestcontainersContainer`.
- We do not enter the semaphore in `CopyFileAsync`, this won’t deadlock KafkaTestcontainer.
    - _I guess we can remove the semaphore from `ExecAsync` too. Multiple threads should be able to copy and execute._
- We create a new GitHub issue that describes the deadlock situation and we fix it in a new pull request.
    - _E. g. we could change the interface in `WithStartupCallback`, to a interface that cannot call methodes that enter the semaphore in `TestcontainersContainer` like `IRunningDockerContainer` or whatever._

I hope I don’t bother you too much. Have a nice day.

Reply:
I have added the semaphore again to `TestcontainersContainer` and removed its usage in `CopyFileAsync` and `ExecAsync`. Also I have moved the configuration code to `KafkaTestcontainerConfiguration`.